### PR TITLE
Added test to verify no duplicate entries

### LINKED ISSUE #1004: Create integration test that checks for duplicate entries after feed update
# Vulnerability detector 

## VUL001

**Short description**

Vulnerability-detector must ignore duplicate entries when updating the database.

**Category**

Vulnerability-detector

**Subcategory**

Database updates

**Description**

Vulnerability-detector should discard any new entry that already exists in a table while an update is taking place. The easiest way to test this out:

1. Create a custom feed and enable the offline update.
2. Start `wazuh-modulesd` an let it update the database.
3. Stop wazuh-modulesd and count the number of entries per table.
4. Modify the feed's timestamp/hash in `METADATA`.
5. Start `wazuh-modulesd` an let it update the database.
6. Stop wazuh-modulesd and count the number of entries per table.

Compare the number of entries between the ones from point 3 and 6. Both should be identical.

**Min/Max compatible versions**
4.0 - Current

**Expected logs**

We should see the usual update logs:

> 2019/09/25 11:31:56 wazuh-modulesd:vulnerability-detector: INFO: (5461): Starting Ubuntu Bionic database update.
> 2019/09/25 11:32:03 wazuh-modulesd:vulnerability-detector: INFO: (5494): The update of the Ubuntu Bionic feed finished successfully.
> 2019/09/25 11:32:03 wazuh-modulesd:vulnerability-detector: INFO: (5461): Starting Ubuntu Xenial database update.
> 2019/09/25 11:32:14 wazuh-modulesd:vulnerability-detector: INFO: (5494): The update of the Ubuntu Xenial feed finished successfully.
...

### PR DESCRIPTION
**Pull request by danimegar**
Hello team,

According to the development in wazuh/wazuh#7050, the RedHat's vulnerabilities are deleted before upgrading the database to avoid duplicate vulnerabilities. This test verifies this development works correctly.

Closes #1004

# Tests logic

`test_scan_results/test_redhat_duplicate_vulns.py`

# Tests checks

- [x] Proven that tests have the expected behavior in DEB**
- [x] Tests pass when using Wazuh v4.0.4 or higher
- [x] Tests should not pass when using a version lower than v4.0.4
- [x] Tests pass in Jenkins: [centOS](https://ci.wazuh.info/job/test_integration/1283/) and [Ubuntu](https://ci.wazuh.info/job/test_integration/1284/)

### Pass expected: v4.0.4
```
# python3 -m pytest -v test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py 
====================================== test session starts ======================================
platform linux -- Python 3.8.5, pytest-6.2.2, py-1.10.0, pluggy-0.13.1 -- /usr/bin/python3
cachedir: .pytest_cache
metadata: {'Python': '3.8.5', 'Platform': 'Linux-5.8.0-41-generic-x86_64-with-glibc2.29', 'Packages': {'pytest': '6.2.2', 'py': '1.10.0', 'pluggy': '0.13.1'}, 'Plugins': {'metadata': '1.11.0', 'html': '2.0.1', 'testinfra': '5.0.0'}}
rootdir: /home/daniel/Wazuh/wazuh-qa/tests/integration, configfile: pytest.ini
plugins: metadata-1.11.0, html-2.0.1, testinfra-5.0.0
collected 3 items                                                                               

test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py::test_redhat_duplicate_vulns[RedHat_6_feed] PASSED [ 33%]
test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py::test_redhat_duplicate_vulns[RedHat_7_feed] PASSED [ 66%]
test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py::test_redhat_duplicate_vulns[RedHat_8_feed] PASSED [100%]

=============================================================== 3 passed in 24.40s ===============================================================

```
### Fail expected: v4.0.3
```
# python3 -m pytest -v test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py 
==================================== test session starts =====================================
platform linux -- Python 3.8.5, pytest-6.2.2, py-1.10.0, pluggy-0.13.1 -- /usr/bin/python3
cachedir: .pytest_cache
metadata: {'Python': '3.8.5', 'Platform': 'Linux-5.8.0-41-generic-x86_64-with-glibc2.29', 'Packages': {'pytest': '6.2.2', 'py': '1.10.0', 'pluggy': '0.13.1'}, 'Plugins': {'metadata': '1.11.0', 'html': '2.0.1', 'testinfra': '5.0.0'}}
rootdir: /home/daniel/Wazuh/wazuh-qa/tests/integration, configfile: pytest.ini
plugins: metadata-1.11.0, html-2.0.1, testinfra-5.0.0
collected 3 items                                                                            

test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py::test_redhat_duplicate_vulns[RedHat_6_feed] FAILED         [ 33%]
test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py::test_redhat_duplicate_vulns[RedHat_7_feed] FAILED         [ 66%]
test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py::test_redhat_duplicate_vulns[RedHat_8_feed] FAILED         [100%]

================================================================== FAILURES ===================================================================
================================================================== FAILURES ===================================================================
_________________________________________________ test_redhat_duplicate_vulns[RedHat_6_feed] __________________________________________________

clean_vuln_tables = None
get_configuration = {'apply_to_modules': ['test_redhat_duplicate_vulns'], 'metadata': {'feed': 'RHEL6'}, 'sections': [{'elements': [{'enab...tributes': [...], 'elements': [...]}}], 'section': 'vulnerability-detector'}], 'tags': ['test_redhat_duplicate_vulns']}
configure_environment = None, restart_modulesd = None

    def test_redhat_duplicate_vulns(clean_vuln_tables, get_configuration, configure_environment,
                                    restart_modulesd):
        """
        RedHat provider was duplicating vulnerabilities when it downloaded a feed to update the database.
        This test check the vulnerabilites are not repeated in the database when it is update.
        """
        feed = get_configuration['metadata']['feed']
        timestamp = '2020-10-31T20:46:48'
    
        vd.check_log_event(wazuh_log_monitor=wazuh_log_monitor,
                           log_event=f"Starting preparse step of feed '{feed}'",
                           update_position=True)
    
        vd.check_log_event(wazuh_log_monitor=wazuh_log_monitor,
                           log_event=r"The update of the 'JSON Red Hat Enterprise Linux' feed finished successfully.",
                           update_position=True)
    
        number_vulns_first = vd.get_num_vulnerabilities()
    
        # Modify feed metadata to download again the feed and update the database
        vd.modify_metadata_vuldet_feed(feed, timestamp)
    
        vd.check_log_event(wazuh_log_monitor=wazuh_log_monitor,
                           log_event=f"Starting preparse step of feed '{feed}'",
                           update_position=False)
    
        vd.check_log_event(wazuh_log_monitor=wazuh_log_monitor,
                           log_event=fr"The update of the 'JSON Red Hat Enterprise Linux' feed finished successfully.",
                           update_position=False)
    
        number_vulns_second = vd.get_num_vulnerabilities()
    
>       assert number_vulns_first == number_vulns_second
E       assert 5 == 10
E         +5
E         -10

```
Best regards.

### CODE REVIEW

File: tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+# Created by Wazuh, Inc. <info@wazuh.com>.
+# This program is free software; you can redistribute it and/or modify it under the terms of GPLv2
+
+import pytest
+import os
+import wazuh_testing.vulnerability_detector as vd
+
+from wazuh_testing.tools import LOG_FILE_PATH
+from wazuh_testing.tools.configuration import load_wazuh_configurations
+from wazuh_testing.tools.monitoring import FileMonitor
```
Comment (BraulioV) [inline tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py:11]:
```suggestion
import os

import pytest
import wazuh_testing.vulnerability_detector as vd
from wazuh_testing.tools import LOG_FILE_PATH
from wazuh_testing.tools.configuration import load_wazuh_configurations
from wazuh_testing.tools.monitoring import FileMonitor
```

File: tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+custom_redhat_json_feed_path = os.path.join(test_data_json, 'feeds', vd.CUSTOM_REDHAT_JSON_FEED)
+
+wazuh_log_monitor = FileMonitor(LOG_FILE_PATH)
+
+# Red Hat configurations
+redhat_configurations = [{'REDHAT_CUSTOM_JSON_FEED': custom_redhat_json_feed_path + '$', 'OS': version,
+                          'REDHAT_CUSTOM_OVAL_FEED': custom_redhat_oval_feed_path}
+                         for version in range(6, 9)]
+redhat_metadata = [{'feed': f'RHEL{version}'}
+                   for version in range(6, 9)]
```
Comment (BraulioV) [inline tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py:31]:
```suggestion
redhat_metadata = [{'feed': f'RHEL{version}'} for version in range(6, 9)]
```

File: tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+
+@pytest.fixture(scope='module', params=configurations, ids=ids)
+def get_configuration(request):
+    """Get configurations from the module."""
+    return request.param
+
+
+def test_redhat_duplicate_vulns(clean_vuln_tables, get_configuration, configure_environment,
+                                restart_modulesd):
```
Comment (BraulioV) [inline tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py:50]:
```suggestion
def test_redhat_duplicate_vulns(clean_vuln_tables, get_configuration, configure_environment, restart_modulesd):
```

**Review by BraulioV — CHANGES_REQUESTED** [review]

File: deps/wazuh_testing/wazuh_testing/vulnerability_detector.py [hunk 4]
```diff
@@ -681,6 +681,11 @@ def check_feed_uncompressed_successfully(wazuh_log_monitor, feed, update_positio
     )
 
 
+def modify_metadata_vuldet_feed(feed, timestamp):
```
Comment (jmv74211) [inline deps/wazuh_testing/wazuh_testing/vulnerability_detector.py:684]: Please, add function documentation in google docstring format. You can find an example here https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

**Review by jmv74211 — DISMISSED** [review]
Good job.

From now on we must place more emphasis on the test and code documentation, since we are working on new repo documentation, and all information about the content is quite helpful.

**Review by BraulioV — APPROVED** [review]

File: deps/wazuh_testing/wazuh_testing/vulnerability_detector.py [hunk 5]
```diff
@@ -681,6 +681,17 @@ def check_feed_uncompressed_successfully(wazuh_log_monitor, feed, update_positio
...
+    """Function to modify the timestamp value of the metadata table for a specific feed.
+
+    Args:
+        feed (str): The feed name.
+        timestamp (str): The new timestamp value to set.
+    """
+    query_string = f'update METADATA set TIMESTAMP="{timestamp}" where TARGET="{feed}"'
```
Comment (jmv74211) [inline deps/wazuh_testing/wazuh_testing/vulnerability_detector.py:691]: Be careful with this, for now whenever you make a query to the database you have to stop `wazuh-db` and start it afterwards.

As a future improvement, we have to delegate this to the make_query function, but for now, you have to do it like this (as it is done in the rest of vuln detector tests)

File: tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+    vd.check_log_event(wazuh_log_monitor=wazuh_log_monitor,
+                       log_event=fr"The update of the 'JSON Red Hat Enterprise Linux' feed finished successfully.",
+                       update_position=False)
+
+    number_vulns_second = vd.get_num_vulnerabilities()
+
+    assert number_vulns_first == number_vulns_second
+
+    vd.clean_vd_tables()
```
Comment (jmv74211) [inline tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py:82]: Care with this, when cleaning tables it is necessary to stop `wazuh-db` daemon. There is a function that can be useful for this case. It is https://github.com/wazuh/wazuh-qa/blob/master/deps/wazuh_testing/wazuh_testing/vulnerability_detector.py#L402

File: tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+    # Modify feed metadata to download again the feed and update the database
+    vd.modify_metadata_vuldet_feed(feed, timestamp)
+
+    vd.check_log_event(wazuh_log_monitor=wazuh_log_monitor,
+                       log_event=f"Starting preparse step of feed '{feed}'",
+                       update_position=False)
+
+    vd.check_log_event(wazuh_log_monitor=wazuh_log_monitor,
+                       log_event=fr"The update of the 'JSON Red Hat Enterprise Linux' feed finished successfully.",
```
Comment (jmv74211) [inline tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py:75]: `fr` flags are not necessary

File: tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    """
+    feed = get_configuration['metadata']['feed']
+    timestamp = '2020-10-31T20:46:48'
+
+    vd.check_log_event(wazuh_log_monitor=wazuh_log_monitor,
+                       log_event=f"Starting preparse step of feed '{feed}'",
+                       update_position=True)
+
+    vd.check_log_event(wazuh_log_monitor=wazuh_log_monitor,
+                       log_event=r"The update of the 'JSON Red Hat Enterprise Linux' feed finished successfully.",
```
Comment (jmv74211) [inline tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py:62]: `r` flag is not necessary

File: tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+
+@pytest.fixture(scope='module', params=configurations, ids=ids)
+def get_configuration(request):
+    """Get configurations from the module."""
+    return request.param
+
+
+def test_redhat_duplicate_vulns(clean_vuln_tables, get_configuration,
+configure_environment, restart_modulesd):
```
Comment (jmv74211) [inline tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py:49]: This line break is not necessary

File: tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    """
+    RedHat provider was duplicating vulnerabilities when it downloaded a feed to update the database.
+    This test check the vulnerabilites are not repeated in the database when it is update.
+    """
+    feed = get_configuration['metadata']['feed']
+    timestamp = '2020-10-31T20:46:48'
+
+    vd.check_log_event(wazuh_log_monitor=wazuh_log_monitor,
+                       log_event=f"Starting preparse step of feed '{feed}'",
+                       update_position=True)
```
Comment (jmv74211) [inline tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py:59]: `update_position=True` for any specific reason?

**Review by jmv74211 — CHANGES_REQUESTED** [review]
Good job. Please review the proposed questions

File: tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py (hunk 10 above)
Comment (danimegar) [reply tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py:58]: Yes, when update_positition is set to False, the test does not respect the interval time and it ends early. The three tests insert vulnerabilities in the VD database at two different times but only one time when update_positition = False.

File: deps/wazuh_testing/wazuh_testing/vulnerability_detector.py (hunk 5 above)
Comment (danimegar) [reply deps/wazuh_testing/wazuh_testing/vulnerability_detector.py:691]: Fixed here: 8c08c1be28917fd543839d17dabc0dc722406080

File: tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py (hunk 6 above)
Comment (danimegar) [reply tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py]: Fixed here: 8c08c1be28917fd543839d17dabc0dc722406080

File: tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py (hunk 7 above)
Comment (danimegar) [reply tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py]: Fixed here: 8c08c1be28917fd543839d17dabc0dc722406080

File: tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py (hunk 8 above)
Comment (danimegar) [reply tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py]: Fixed here: 8c08c1be28917fd543839d17dabc0dc722406080

File: tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py (hunk 9 above)
Comment (danimegar) [reply tests/integration/test_vulnerability_detector/test_scan_results/test_redhat_duplicate_vulns.py]: Fixed here: 8c08c1be28917fd543839d17dabc0dc722406080

**Review by chemamartinez — APPROVED** [review]

**Review by jmv74211 — APPROVED** [review]

**Review by BraulioV — APPROVED** [review]

**Pull request merged**